src/disk_cache.rs

```diff
@@ -0,0 +1,1064 @@
+                            source: Box::new(InvalidGetRange::StartTooLarge {
+                                requested: range.start,
+                                length: object_size,
+                            }),
+                        });
+                    }
+                    if range.start >= range.end {
+                        return Err(object_store::Error::Generic {
+                            store: "cached_object_store",
+                            source: Box::new(InvalidGetRange::Inconsistent {
+                                start: range.start,
+                                end: range.end,
+                            }),
+                        });
+                    }
+                    (range.start, range.end.min(object_size))
+                }
+                GetRange::Offset(offset) => {
+                    if offset >= object_size {
+                        return Err(object_store::Error::Generic {
+                            store: "cached_object_store",
+                            source: Box::new(InvalidGetRange::StartTooLarge {
+                                requested: offset,
+                                length: object_size,
+                            }),
+                        });
+                    }
+                    (offset, object_size)
+                }
+                GetRange::Suffix(suffix) => (object_size.saturating_sub(suffix), object_size),
+            },
+        };
+        Ok(Range {
+            start: start_offset,
+            end: end_offset,
+        })
+    }
+
+    fn align_get_range(&self, range: &GetRange) -> GetRange {
+        match range {
+            GetRange::Bounded(bounded) => {
+                let aligned = self.align_range(bounded, self.part_size);
+                GetRange::Bounded(aligned)
+            }
+            GetRange::Suffix(suffix) => {
+                let suffix_aligned = self.align_range(&(0..*suffix), self.part_size).end;
+                GetRange::Suffix(suffix_aligned)
+            }
+            GetRange::Offset(offset) => {
+                let offset_aligned = *offset - *offset % self.part_size;
+                GetRange::Offset(offset_aligned)
+            }
+        }
+    }
+
+    fn align_range(&self, range: &Range<usize>, alignment: usize) -> Range<usize> {
+        let start_aligned = range.start - range.start % alignment;
+        let end_aligned = ((range.end + alignment - 1) / alignment) * alignment;
+        Range {
+            start: start_aligned,
+            end: end_aligned,
+        }
+    }
+}
+
+impl std::fmt::Display for CacheableObjectStoreInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "LocalCachedObjectStore({}, {})",
+            self.object_store, self.storage
+        )
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct LocalCacheHead {
+    pub location: String,
```

Comment:
Head would be very small, probably 1kb. Should head be cached in memory as well?
1000 files should take few MBs.

Reply:
i think linux's page cache may help us this a little bit. 

in theory, all these on disk data could be warmed with page cache in memory, and the page cache keeps warm even after process restart.